packages/wonder-blocks-modal/components/modal-backdrop.js

```diff
@@ -70,12 +70,147 @@ export default class ModalBackdrop extends React.Component<Props> {
+                  * even if the modal is mounted at the start/end of the
+                  * document (end is likely in practice!), we add focusable
+                  * sentinel nodes. That way, tabbing out of the modal is more
+                  * likely to hit a sentinel node. They're not critical to
+                  * focus wrapping, though; we're resilient to any kind of focus
+                  * shift, whether they hit the sentinels or not! */}
+                <div tabIndex="0" />
+                <FocusTrap>{clonedChildren}</FocusTrap>
+                <div tabIndex="0" />
             </View>
         );
     }
 }
 
+/**
+ * This component ensures that focus stays within itself. If the user uses Tab
+ * at the end of the modal, or Shift-Tab at the start of the modal, then this
+ * component wraps focus to the start/end respectively.
+ *
+ * Adapted from the WAI-ARIA dialog behavior example.
+ * https://www.w3.org/TR/2017/NOTE-wai-aria-practices-1.1-20171214/examples/dialog-modal/dialog.html
+ */
+class FocusTrap extends React.Component<{children: React.Node}> {
+    /** The most recent node _inside this component_ to receive focus. */
+    _lastNodeFocusedInModal: ?Node = null;
+
+    /**
+     * Whether we're currently applying programmatic focus, and should therefore
+     * ignore focus change events.
+     */
+    _ignoreFocusChanges: boolean = false;
+
+    componentDidMount() {
+        window.addEventListener("focus", this._handleGlobalFocus, true);
+    }
+
+    componentWillUnmount() {
+        window.removeEventListener("focus", this._handleGlobalFocus, true);
+    }
+
+    /** Get the outermost DOM node of our children. */
+    _getModalRoot(): Node {
+        return (ReactDOM.findDOMNode(this): any);
+    }
+
+    /** Try to focus the given node. Return true iff successful. */
+    _tryToFocus(node: Node) {
+        if (typeof node.focus !== "function") {
+            return;
+        }
+
+        this._ignoreFocusChanges = true;
+        try {
+            node.focus();
+        } catch (e) {}
+        this._ignoreFocusChanges = false;
+
+        return document.activeElement === node;
+    }
+
+    /**
+     * Focus the first focusable descendant of the given node.
+     *
+     * Return true if we succeed. Or, if the given node has no focusable
+     * descendants, return false.
+     */
+    _focusFirstElementIn(currentParent: Node) {
+        const children = currentParent.childNodes;
+        for (let i = 0; i < children.length; i++) {
+            const child = children[i];
+            if (this._tryToFocus(child) || this._focusFirstElementIn(child)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Focus the last focusable descendant of the given node.
+     *
+     * Return true if we succeed. Or, if the given node has no focusable
+     * descendants, return false.
+     */
+    _focusLastElementIn(currentParent: Node) {
```

Comment:
Hehe yeah, I was gonna try something like that! Might pull this focus behavior into util functions so that we can go even more granular—something like "focus the first element in the content area, or, if unsuccessful, focus the last element in the footer area". Fingers crossed that it's just easy!

Reply:
(If so, though, I'll do that extraction during WB-151, rather than this change 👍🏻)